File: core/src/main/java/org/iumotionlabs/hefesto/feature/procwatch/model/AlertRule.java
package org.iumotionlabs.hefesto.feature.procwatch.model;

import java.time.Duration;
import java.util.Optional;

/**
 * Record representing an alert rule with condition and optional time window.
 */
public record AlertRule(
    String expression,
    MetricType metric,
    Operator operator,
    double threshold,
    ThresholdUnit unit,
    Optional<Duration> window,
    Optional<WindowCondition> windowCondition
) {
    /**
     * Metrics that can be monitored.
     */
    public enum MetricType {
        CPU("cpu", "CPU percentage"),
        RSS("rss", "Resident Set Size (memory)"),
        VIRTUAL("virtual", "Virtual memory"),
        THREADS("threads", "Thread count"),
        FD("fd", "File descriptors"),
        READ_BYTES("read", "Read bytes"),
        WRITE_BYTES("write", "Write bytes");

        private final String name;
        private final String description;

        MetricType(String name, String description) {
            this.name = name;
            this.description = description;
        }

        public String getName() {
            return name;
        }

        public String getDescription() {
            return description;
        }

        public static MetricType fromString(String s) {
            String lower = s.toLowerCase();
            return switch (lower) {
                case "cpu", "cpu%" -> CPU;
                case "rss", "mem", "memory" -> RSS;
                case "virtual", "vsz", "virt" -> VIRTUAL;
                case "threads", "thread" -> THREADS;
                case "fd", "fds", "files" -> FD;
                case "read", "read_bytes" -> READ_BYTES;
                case "write", "write_bytes" -> WRITE_BYTES;
                default -> throw new IllegalArgumentException("Unknown metric: " + s);
            };
        }
    }

    /**
     * Comparison operators.
     */
    public enum Operator {
        GREATER(">"),
        GREATER_EQ(">="),
        LESS("<"),
        LESS_EQ("<="),
        EQUALS("=="),
        NOT_EQUALS("!=");

        private final String symbol;

        Operator(String symbol) {
            this.symbol = symbol;
        }

        public String symbol() {
            return symbol;
        }

        public boolean evaluate(double value, double threshold) {
            return switch (this) {
                case GREATER -> value > threshold;
                case GREATER_EQ -> value >= threshold;
                case LESS -> value < threshold;
                case LESS_EQ -> value <= threshold;
                case EQUALS -> Math.abs(value - threshold) < 0.001;
                case NOT_EQUALS -> Math.abs(value - threshold) >= 0.001;
            };
        }

        public static Operator fromString(String s) {
            return switch (s) {
                case ">" -> GREATER;
                case ">=" -> GREATER_EQ;
                case "<" -> LESS;
                case "<=" -> LESS_EQ;
                case "==", "=" -> EQUALS;
                case "!=", "<>" -> NOT_EQUALS;
                default -> throw new IllegalArgumentException("Unknown operator: " + s);
            };
        }
    }

    /**
     * Unit for threshold values.
     */
    public enum ThresholdUnit {
        NONE(1),
        PERCENT(1),
        BYTES(1),
        KB(1024),
        MB(1024 * 1024),
        GB(1024L * 1024 * 1024);

        private final long multiplier;

        ThresholdUnit(long multiplier) {
            this.multiplier = multiplier;
        }

        public long multiplier() {
            return multiplier;
        }

        public double toBytes(double value) {
            return value * multiplier;
        }

        public static ThresholdUnit fromString(String s) {
            if (s == null || s.isEmpty()) return NONE;
            return switch (s.toUpperCase()) {
                case "%" -> PERCENT;
                case "B", "BYTES" -> BYTES;
                case "K", "KB" -> KB;
                case "M", "MB" -> MB;
                case "G", "GB" -> GB;
                default -> NONE;
            };
        }
    }

    /**
     * Window conditions for time-based alerts.
     */
    public enum WindowCondition {
        FOR("for"),           // Condition must hold for duration
        INCREASING("increasing"), // Value increasing over duration
        DECREASING("decreasing"); // Value decreasing over duration

        private final String keyword;

        WindowCondition(String keyword) {
            this.keyword = keyword;
        }

        public String keyword() {
            return keyword;
        }

        public static WindowCondition fromString(String s) {
            return switch (s.toLowerCase()) {
                case "for" -> FOR;
                case "increasing" -> INCREASING;
                case "decreasing" -> DECREASING;
                default -> throw new IllegalArgumentException("Unknown window condition: " + s);
            };
        }
    }

    /**
     * Evaluates this rule against a process sample.
     */
    public boolean evaluate(ProcessSample sample) {
        double value = extractMetricValue(sample);
        double normalizedThreshold = normalizeThreshold();
        return operator.evaluate(value, normalizedThreshold);
    }

    /**
     * Extracts the metric value from a sample.
     */
    public double extractMetricValue(ProcessSample sample) {
        return switch (metric) {
            case CPU -> sample.cpu().percentInstant();
            case RSS -> sample.memory().rssBytes();
            case VIRTUAL -> sample.memory().virtualBytes();
            case THREADS -> sample.threadCount();
            case FD -> sample.openFileDescriptors();
            case READ_BYTES -> sample.io().readBytes();
            case WRITE_BYTES -> sample.io().writeBytes();
        };
    }

    /**
     * Normalizes threshold based on unit.
     */
    public double normalizeThreshold() {
        if (metric == MetricType.CPU || metric == MetricType.THREADS || metric == MetricType.FD) {
            return threshold;
        }
        return unit.toBytes(threshold);
    }

    /**
     * Returns a human-readable description.
     */
    public String describe() {
        StringBuilder sb = new StringBuilder();
        sb.append(metric.getName());
        sb.append(" ").append(operator.symbol());
        sb.append(" ").append(formatThreshold());

        window.ifPresent(w -> {
            sb.append(" ");
            windowCondition.ifPresent(c -> sb.append(c.keyword()).append(" "));
            sb.append(formatDuration(w));
        });

        return sb.toString();
    }

    private String formatThreshold() {
        if (unit == ThresholdUnit.PERCENT) {
            return "%.0f%%".formatted(threshold);
        } else if (unit == ThresholdUnit.GB) {
            return "%.1fGB".formatted(threshold);
        } else if (unit == ThresholdUnit.MB) {
            return "%.0fMB".formatted(threshold);
        } else if (unit == ThresholdUnit.KB) {
            return "%.0fKB".formatted(threshold);
        } else if (threshold == (long) threshold) {
            return String.valueOf((long) threshold);
        } else {
            return String.valueOf(threshold);
        }
    }

    private String formatDuration(Duration d) {
        long seconds = d.getSeconds();
        if (seconds < 60) return seconds + "s";
        if (seconds < 3600) return (seconds / 60) + "m";
        return (seconds / 3600) + "h";
    }

    /**
     * Creates a simple threshold rule.
     */
    public static AlertRule simple(MetricType metric, Operator op, double threshold, ThresholdUnit unit) {
        String expr = metric.getName() + op.symbol() + threshold + (unit != ThresholdUnit.NONE ? unit.name() : "");
        return new AlertRule(expr, metric, op, threshold, unit, Optional.empty(), Optional.empty());
    }

    /**
     * Creates a windowed rule.
     */
    public static AlertRule windowed(MetricType metric, Operator op, double threshold, ThresholdUnit unit,
                                     Duration window, WindowCondition condition) {
        String expr = metric.getName() + op.symbol() + threshold +
                     (unit != ThresholdUnit.NONE ? unit.name() : "") +
                     " " + condition.keyword() + " " + window.getSeconds() + "s";
        return new AlertRule(expr, metric, op, threshold, unit, Optional.of(window), Optional.of(condition));
    }
}

File: hefesto-domain/src/procwatch/alert_rule.rs
use serde::Serialize;
use std::time::Duration;
use thiserror::Error;

use super::process_sample::ProcessSample;

#[derive(Debug, Error)]
pub enum AlertRuleError {
    #[error("Unknown metric: {0}")]
    UnknownMetric(String),
    #[error("Unknown operator: {0}")]
    UnknownOperator(String),
    #[error("Unknown window condition: {0}")]
    UnknownWindowCondition(String),
}

/// Metrics that can be monitored.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub enum MetricType {
    Cpu,
    Rss,
    Virtual,
    Threads,
    Fd,
    ReadBytes,
    WriteBytes,
}

impl MetricType {
    pub fn name(&self) -> &str {
        match self {
            MetricType::Cpu => "cpu",
            MetricType::Rss => "rss",
            MetricType::Virtual => "virtual",
            MetricType::Threads => "threads",
            MetricType::Fd => "fd",
            MetricType::ReadBytes => "read",
            MetricType::WriteBytes => "write",
        }
    }

    pub fn description(&self) -> &str {
        match self {
            MetricType::Cpu => "CPU percentage",
            MetricType::Rss => "Resident Set Size (memory)",
            MetricType::Virtual => "Virtual memory",
            MetricType::Threads => "Thread count",
            MetricType::Fd => "File descriptors",
            MetricType::ReadBytes => "Read bytes",
            MetricType::WriteBytes => "Write bytes",
        }
    }

    pub fn parse(s: &str) -> Result<Self, AlertRuleError> {
        match s.to_lowercase().as_str() {
            "cpu" | "cpu%" => Ok(MetricType::Cpu),
            "rss" | "mem" | "memory" => Ok(MetricType::Rss),
            "virtual" | "vsz" | "virt" => Ok(MetricType::Virtual),
            "threads" | "thread" => Ok(MetricType::Threads),
            "fd" | "fds" | "files" => Ok(MetricType::Fd),
            "read" | "read_bytes" => Ok(MetricType::ReadBytes),
            "write" | "write_bytes" => Ok(MetricType::WriteBytes),
            other => Err(AlertRuleError::UnknownMetric(other.to_string())),
        }
    }
}

/// Comparison operators.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub enum ComparisonOperator {
    Greater,
    GreaterEq,
    Less,
    LessEq,
    Equals,
    NotEquals,
}

impl ComparisonOperator {
    pub fn symbol(&self) -> &str {
        match self {
            ComparisonOperator::Greater => ">",
            ComparisonOperator::GreaterEq => ">=",
            ComparisonOperator::Less => "<",
            ComparisonOperator::LessEq => "<=",
            ComparisonOperator::Equals => "==",
            ComparisonOperator::NotEquals => "!=",
        }
    }

    pub fn evaluate(&self, value: f64, threshold: f64) -> bool {
        match self {
            ComparisonOperator::Greater => value > threshold,
            ComparisonOperator::GreaterEq => value >= threshold,
            ComparisonOperator::Less => value < threshold,
            ComparisonOperator::LessEq => value <= threshold,
            ComparisonOperator::Equals => (value - threshold).abs() < 0.001,
            ComparisonOperator::NotEquals => (value - threshold).abs() >= 0.001,
        }
    }

    pub fn parse(s: &str) -> Result<Self, AlertRuleError> {
        match s {
            ">" => Ok(ComparisonOperator::Greater),
            ">=" => Ok(ComparisonOperator::GreaterEq),
            "<" => Ok(ComparisonOperator::Less),
            "<=" => Ok(ComparisonOperator::LessEq),
            "==" | "=" => Ok(ComparisonOperator::Equals),
            "!=" | "<>" => Ok(ComparisonOperator::NotEquals),
            other => Err(AlertRuleError::UnknownOperator(other.to_string())),
        }
    }
}

impl std::fmt::Display for ComparisonOperator {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.symbol())
    }
}

/// Unit for threshold values.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub enum ThresholdUnit {
    None,
    Percent,
    Bytes,
    Kb,
    Mb,
    Gb,
}

impl ThresholdUnit {
    pub fn multiplier(&self) -> u64 {
        match self {
            ThresholdUnit::None | ThresholdUnit::Percent | ThresholdUnit::Bytes => 1,
            ThresholdUnit::Kb => 1024,
            ThresholdUnit::Mb => 1024 * 1024,
            ThresholdUnit::Gb => 1024 * 1024 * 1024,
        }
    }

    pub fn to_bytes(&self, value: f64) -> f64 {
        value * self.multiplier() as f64
    }

    pub fn parse(s: &str) -> Self {
        if s.is_empty() {
            return ThresholdUnit::None;
        }
        match s.to_uppercase().as_str() {
            "%" => ThresholdUnit::Percent,
            "B" | "BYTES" => ThresholdUnit::Bytes,
            "K" | "KB" => ThresholdUnit::Kb,
            "M" | "MB" => ThresholdUnit::Mb,
            "G" | "GB" => ThresholdUnit::Gb,
            _ => ThresholdUnit::None,
        }
    }
}

/// Window conditions for time-based alerts.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub enum WindowCondition {
    For,
    Increasing,
    Decreasing,
}

impl WindowCondition {
    pub fn keyword(&self) -> &str {
        match self {
            WindowCondition::For => "for",
            WindowCondition::Increasing => "increasing",
            WindowCondition::Decreasing => "decreasing",
        }
    }

    pub fn parse(s: &str) -> Result<Self, AlertRuleError> {
        match s.to_lowercase().as_str() {
            "for" => Ok(WindowCondition::For),
            "increasing" => Ok(WindowCondition::Increasing),
            "decreasing" => Ok(WindowCondition::Decreasing),
            other => Err(AlertRuleError::UnknownWindowCondition(other.to_string())),
        }
    }
}

/// An alert rule with condition and optional time window.
#[derive(Debug, Clone, Serialize)]
pub struct AlertRule {
    pub expression: String,
    pub metric: MetricType,
    pub operator: ComparisonOperator,
    pub threshold: f64,
    pub unit: ThresholdUnit,
    pub window: Option<Duration>,
    pub window_condition: Option<WindowCondition>,
}

impl AlertRule {
    /// Evaluates this rule against a process sample.
    pub fn evaluate(&self, sample: &ProcessSample) -> bool {
        let value = self.extract_metric_value(sample);
        let normalized_threshold = self.normalize_threshold();
        self.operator.evaluate(value, normalized_threshold)
    }

    /// Extracts the metric value from a sample.
    pub fn extract_metric_value(&self, sample: &ProcessSample) -> f64 {
        match self.metric {
            MetricType::Cpu => sample.cpu.percent_instant,
            MetricType::Rss => sample.memory.rss_bytes as f64,
            MetricType::Virtual => sample.memory.virtual_bytes as f64,
            MetricType::Threads => sample.thread_count as f64,
            MetricType::Fd => sample.open_file_descriptors as f64,
            MetricType::ReadBytes => sample.io.read_bytes as f64,
            MetricType::WriteBytes => sample.io.write_bytes as f64,
        }
    }

    /// Normalizes threshold based on unit.
    pub fn normalize_threshold(&self) -> f64 {
        match self.metric {
            MetricType::Cpu | MetricType::Threads | MetricType::Fd => self.threshold,
            _ => self.unit.to_bytes(self.threshold),
        }
    }

    /// Returns a human-readable description.
    pub fn describe(&self) -> String {
        let mut sb = String::new();
        sb.push_str(self.metric.name());
        sb.push(' ');
        sb.push_str(self.operator.symbol());
        sb.push(' ');
        sb.push_str(&self.format_threshold());

        if let Some(ref w) = self.window {
            sb.push(' ');
            if let Some(ref wc) = self.window_condition {
                sb.push_str(wc.keyword());
                sb.push(' ');
            }
            sb.push_str(&format_duration(w));
        }

        sb
    }

    fn format_threshold(&self) -> String {
        match self.unit {
            ThresholdUnit::Percent => format!("{:.0}%", self.threshold),
            ThresholdUnit::Gb => format!("{:.1}GB", self.threshold),
            ThresholdUnit::Mb => format!("{:.0}MB", self.threshold),
            ThresholdUnit::Kb => format!("{:.0}KB", self.threshold),
            _ => {
                if self.threshold == (self.threshold as i64) as f64 {
                    format!("{}", self.threshold as i64)
                } else {
                    format!("{}", self.threshold)
                }
            }
        }
    }

    /// Creates a simple threshold rule.
    pub fn simple(
        metric: MetricType,
        operator: ComparisonOperator,
        threshold: f64,
        unit: ThresholdUnit,
    ) -> Self {
        let unit_str = match unit {
            ThresholdUnit::None => String::new(),
            ThresholdUnit::Percent => "%".to_string(),
            ThresholdUnit::Bytes => "B".to_string(),
            ThresholdUnit::Kb => "KB".to_string(),
            ThresholdUnit::Mb => "MB".to_string(),
            ThresholdUnit::Gb => "GB".to_string(),
        };
        let expr = format!(
            "{}{}{}{}",
            metric.name(),
            operator.symbol(),
            threshold,
            unit_str
        );
        Self {
            expression: expr,
            metric,
            operator,
            threshold,
            unit,
            window: None,
            window_condition: None,
        }
    }

    /// Creates a windowed rule.
    pub fn windowed(
        metric: MetricType,
        operator: ComparisonOperator,
        threshold: f64,
        unit: ThresholdUnit,
        window: Duration,
        condition: WindowCondition,
    ) -> Self {
        let unit_str = match unit {
            ThresholdUnit::None => String::new(),
            ThresholdUnit::Percent => "%".to_string(),
            ThresholdUnit::Bytes => "B".to_string(),
            ThresholdUnit::Kb => "KB".to_string(),
            ThresholdUnit::Mb => "MB".to_string(),
            ThresholdUnit::Gb => "GB".to_string(),
        };
        let expr = format!(
            "{}{}{}{} {} {}s",
            metric.name(),
            operator.symbol(),
            threshold,
            unit_str,
            condition.keyword(),
            window.as_secs()
        );
        Self {
            expression: expr,
            metric,
            operator,
            threshold,
            unit,
            window: Some(window),
            window_condition: Some(condition),
        }
    }
}

fn format_duration(d: &Duration) -> String {
    let seconds = d.as_secs();
    if seconds < 60 {
        format!("{}s", seconds)
    } else if seconds < 3600 {
        format!("{}m", seconds / 60)
    } else {
        format!("{}h", seconds / 3600)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_metric_type_from_str() {
        assert_eq!(MetricType::parse("cpu").unwrap(), MetricType::Cpu);
        assert_eq!(MetricType::parse("CPU%").unwrap(), MetricType::Cpu);
        assert_eq!(MetricType::parse("rss").unwrap(), MetricType::Rss);
        assert_eq!(MetricType::parse("mem").unwrap(), MetricType::Rss);
        assert_eq!(MetricType::parse("memory").unwrap(), MetricType::Rss);
        assert_eq!(MetricType::parse("virtual").unwrap(), MetricType::Virtual);
        assert_eq!(MetricType::parse("threads").unwrap(), MetricType::Threads);
        assert!(MetricType::parse("invalid").is_err());
    }

    #[test]
    fn test_operator_evaluate() {
        assert!(ComparisonOperator::Greater.evaluate(90.0, 80.0));
        assert!(!ComparisonOperator::Greater.evaluate(80.0, 90.0));
        assert!(ComparisonOperator::GreaterEq.evaluate(80.0, 80.0));
        assert!(ComparisonOperator::Less.evaluate(70.0, 80.0));
        assert!(ComparisonOperator::Equals.evaluate(80.0, 80.0));
        assert!(ComparisonOperator::NotEquals.evaluate(70.0, 80.0));
    }

    #[test]
    fn test_simple_rule_evaluate() {
        let rule = AlertRule::simple(
            MetricType::Cpu,
            ComparisonOperator::Greater,
            80.0,
            ThresholdUnit::Percent,
        );

        let mut sample = ProcessSample::minimal(1, "test", "user");
        sample.cpu.percent_instant = 90.0;
        assert!(rule.evaluate(&sample));

        sample.cpu.percent_instant = 70.0;
        assert!(!rule.evaluate(&sample));
    }

    #[test]
    fn test_describe() {
        let rule = AlertRule::simple(
            MetricType::Cpu,
            ComparisonOperator::Greater,
            80.0,
            ThresholdUnit::Percent,
        );
        assert_eq!(rule.describe(), "cpu > 80%");
    }

    #[test]
    fn test_describe_windowed() {
        let rule = AlertRule::windowed(
            MetricType::Cpu,
            ComparisonOperator::Greater,
            80.0,
            ThresholdUnit::Percent,
            Duration::from_secs(30),
            WindowCondition::For,
        );
        assert_eq!(rule.describe(), "cpu > 80% for 30s");
    }

    #[test]
    fn test_threshold_unit_to_bytes() {
        assert_eq!(
            ThresholdUnit::Gb.to_bytes(1.5),
            1.5 * 1024.0 * 1024.0 * 1024.0
        );
        assert_eq!(ThresholdUnit::Mb.to_bytes(100.0), 100.0 * 1024.0 * 1024.0);
        assert_eq!(ThresholdUnit::Kb.to_bytes(512.0), 512.0 * 1024.0);
    }
}
